[1주차] 간단한 계산기

jihwooon opened the pull request:
# 와일드 백엔드 1주차 과제

## 목적
Java Http 통신으로 계산기 API 구현하기

## 배경
계산기를 구현함으로써 질문과 피드백 반영


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **신규 기능**
	- 기본 산술 연산(덧셈, 뺄셈, 곱셈, 나눗셈)을 수행하는 API 추가.
	- 계산 결과를 반환하는 `POST /calculations` 엔드포인트 및 계산 이력을 조회하는 `GET /calculations` 엔드포인트 추가.
	- 계산 목록을 반환하는 새로운 `GET /calculations` 요청 추가.
	- 잘못된 연산 기호에 대한 오류 처리 기능 추가.
- **문서화**
	- README.md에 API 사용법 및 예제 추가.
	- API의 오류 처리 및 동작 방식에 대한 자세한 설명 포함.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jihwooon committed:
사칙연산 계산기 API에 대한 README 업데이트

이 커밋은 사칙연산 계산기 API에 대한 상세한 설명과 API 설계를 README에 추가합니다.

주요 변경사항:
1. 프로젝트 개요와 조건에 대한 설명 추가: Layered Architecture를 기반으로 한 사칙연산 계산기 API에 대한 개요와 조건을 추가하였습니다.
2. API 설계 추가: POST /calculations와 GET /calculations에 대한 요청과 응답 예시를 추가하였습니다.

jihwooon committed:
README 문서의 제목 수준 변경

README 문서의 제목 수준이 변경되었습니다. 이는 문서의 구조를 더 명확하게 표현하기 위한 변경입니다.

주요 변경사항:
1. "과제 개요"의 제목 수준을 ###에서 ##로 변경하였습니다. 이는 해당 섹션의 중요성을 강조하기 위함입니다.

ahastudio commented on README.md:
````diff
@@ -1 +1,73 @@
+사칙 연산(`operations`) 외 다른 문자가 들어오면 `올바른지 않은 사칙연산 문자입니다.` 에러가 발생합니다.
+
+### API 설계
+```json
````
아래 내용을 보면 JSON이 아닙니다. GitHub의 경우 문제가 있는 부분을 아주 명확히 시각화합니다. 뭔가 다르게 보이는 걸 항상 확인해 보세요.

![Screenshot](https://github.com/user-attachments/assets/834b8dc5-5d05-4909-9813-10ad1b5f08e6)

ahastudio commented:
Visual Studio Code, 플러그인 등 기타 도구를 이용해서 마크다운도 Lint할 수 있습니다. 처음에는 생각보다 정말 많은 걸 발견할 수 있습니다.

참고로, 제가 쓰는 건 이겁니다: https://github.com/DavidAnson/markdownlint

ahastudio commented on README.md:
```diff
@@ -1 +1,73 @@
+수행된 모든 연산 기록을 반환합니다. 각 연산 기록에는 두 숫자, 연산 기호, 결과 값이 포함됩니다.
+
+Response:
+[
```
최상위를 array로 잡지 마세요. 확장성이 매우 낮아집니다. 가장 대표적인 게 바로 pagination 지원이죠. totalCount, totalPages, currentPage 등 다양한 정보가 함께 전달돼야 하는데, 최상위가 array인 경우엔 FE와 BE가 일시에 이를 object로 바꿔야 합니다. 하지만 처음부터 object면 전혀 문제가 안 되죠.

jihwooon replied:
매번 array를 최상위로 직접 반영해서 프론트로 넘겼는데 확장성에 불리하다는 것을 이번에 알게 되었습니다.

ahastudio commented:
FE와 협업하는 상황이라면 두 가지를 구분하는 게 좋습니다. 하나는 타입과 설명에 대한 거고, 다른 하나는 예제입니다. 전자는 TypeScript로 맞춰주면 더 좋습니다.

더 좋은 건 OpenAPI 스펙으로 작성하는 겁니다. 이렇게 하면 문서화, 예제 코드 생성, 가짜 서버 제공 등 다양한 이득을 얻을 수 있습니다. https://spec.openapis.org/

jihwooon committed:
README 문서의 구조와 내용 개선

README 문서의 구조를 개선하고, 일부 내용을 명확하게 수정하였습니다.

주요 변경사항:
1. 제목의 계층 구조를 재정리하였습니다. 예를 들어, "조건"과 "API 설계"는 이제 2단계 제목으로 표시됩니다.
2. API 설계 섹션의 내용을 개선하였습니다. 각 API 요청과 응답에 대한 설명이 더욱 명확해졌습니다.
3. 일부 문장을 더 자연스럽게 수정하였습니다. 예를 들어, "사칙연산(`operations`)에 따라 숫자 연산이 주어지면 조건에 맞게 값이 나오게 한다."는 "사칙연산(`operations`)에 따라 숫자 연산이 주어지면 조건에 맞게 값이 나오도록 한다."로 변경되었습니다.

ahastudio commented:
구현 들어가도 좋을 것 같습니다. OpenAPI 스펙에 대한 자료는 따로 전달 드릴게요.

jihwooon committed:
간단한 계산기 API에 대한 OpenAPI 문서 추가

이 커밋은 간단한 계산기 API에 대한 OpenAPI 문서를 추가합니다. 이 문서는 API의 사용 방법을 명확하게 설명하고, API를 사용하는 클라이언트 개발에 도움을 줍니다.

주요 변경사항:
1. OpenAPI 문서 생성: `docs/openapi.yaml` 파일을 새로 생성하였습니다.
2. API 정보 추가: API의 제목, 설명, 버전 등의 정보를 추가하였습니다.
3. API 경로 정의: `/calculations` 경로에 대한 POST와 GET 메소드를 정의하였습니다.
4. 요청 및 응답 스키마 정의: 각 메소드에 대한 요청과 응답의 스키마를 정의하였습니다.

jihwooon commented:
> 구현 들어가도 좋을 것 같습니다. OpenAPI 스펙에 대한 자료는 따로 전달 드릴게요.

말씀 주신 부분 반영해서 이제 구현으로 들어가보겟습니다~!!

jihwooon commented on docs/opeanapi.yaml:
```diff
@@ -0,0 +1,68 @@
+                                number2:
+                                    type: number
+                                    description: 두번째 number 인수
+                                operations:
+                                    type: string
+                                    description: 사칙연산("+", "-", "*", "/")
+                                    example: "+"
```
현재 operations은 enum으로 관리하지 않습니다.

ahastudio replied:
여기서 enum은 코드가 아니라 API 레벨의 문제라, 가능하면 enum으로 하는 게 좋습니다.

jihwooon committed:
서버 URL을 정확하게 업데이트

OpenAPI 문서의 서버 URL이 잘못되어 있어 이를 정확한 URL로 수정하였습니다.

주요 변경사항:
1. 서버 URL이 'https'에서 'https://api.wild-backend.com/v1'로 변경되었습니다.

jihwooon commented on README.md:
````diff
@@ -1 +1,84 @@
+
+```
+
+**GET /calculations**
+
+수행된 모든 연산 기록을 반환합니다. 각 연산 기록에는 두 숫자, 연산 기호, 결과 값이 포함됩니다.
+
+
+**Response:**
+
+```json
+{
+    "calculations": [
+        {
+            "number1": 1,
+            "number2": 2,
+            "operations": "+",
+            "result": 3
+        },
+        {
+            "number1": 3,
+            "number2": 2,
+            "operations": "-",
+            "result": 1
+        },
+        {
+            "number1": 4,
+            "number2": 2,
+            "operations": "*",
+            "result": 8
+        },
+        {
+            "number1": 4,
+            "number2": 2,
+            "operations": "/",
+            "result": 2
+        }
+    ]
+}
+```
````
array 타입을 사용하지만 
  ```
  "totalCount": 2,
  "currentPage": 1,
  "totalPages": 1
  ```
  은 이번 api에 포함하지 않습니다. 그렇지만 이 페이지네이션가 필요 할까요?

ahastudio commented on README.md:
```diff
@@ -1 +1,84 @@
+
+사칙 연산(`operations`) 외 다른 문자가 들어오면 `올바른지 않은 사칙연산 문자입니다.` 에러가 발생합니다.
+
+## API 설계
```
OpenAPI로 스펙 관리할 거면 README의 내용은 삭제하는 게 일관성을 위해 유리합니다.

jihwooon committed:
README 문서에서 API 설계 부분 제거

README 문서에서 API 설계 부분이 제거되었습니다. 이 부분은 이제 더 이상 필요하지 않거나 다른 문서로 이동되었을 수 있습니다.

주요 변경사항:
1. POST /calculations에 대한 설명 제거
2. GET /calculations에 대한 설명 제거
3. 각 API 요청과 응답 예시 제거

jihwooon committed:
HTTP 서버 설정 추가

기존의 "Hello, world!"를 출력하던 App 클래스가 HTTP 서버를 생성하고 시작하는 기능을 가지도록 변경되었습니다.

주요 변경사항:
1. `HttpServer`를 import하고, 서버를 생성하고 시작하는 코드를 추가하였습니다.
2. 서버의 호스트와 포트는 각각 "localhost"와 8080으로 설정하였습니다.
3. 서버가 시작되면 "Listening on http://localhost:8080"이라는 메시지를 출력하도록 하였습니다.

jihwooon committed:
사칙연산 enum 값 추가

OpenAPI 문서에 사칙연산 필드에 대한 enum 값을 추가하였습니다. 이로써 사용자는 사칙연산 필드에 "+", "-", "*", "/" 중 하나의 값을 입력해야 합니다.

주요 변경사항:
1. 사칙연산 필드에 대한 enum 값 추가: 이전에는 사용자가 임의의 문자열을 입력할 수 있었지만, 이제는 "+", "-", "*", "/" 중 하나의 값을 입력해야 합니다.

jihwooon committed:
HTTP 서버에 RequestHandler 추가

HTTP 서버가 클라이언트의 요청을 처리할 수 있도록 RequestHandler를 추가하였습니다.

주요 변경사항:
1. App.java에서 HTTP 서버를 생성할 때, "/" 경로에 대한 요청을 처리하는 RequestHandler를 추가하였습니다.
2. RequestHandler.java 파일을 새로 생성하였고, 이 파일에서는 클라이언트의 요청을 받아 "Hello, World!"를 응답하는 기능을 구현하였습니다.

jihwooon committed:
RequestHandler에서 요청 키 생성 로직 추가

RequestHandler의 handle 메소드가 수정되었습니다. 이전에는 단순히 "Hello, World!"를 응답으로 보내는 기능만 있었지만, 이제는 요청 메소드와 경로를 결합하여 요청 키를 생성하는 기능이 추가되었습니다.

주요 변경사항:
1. "Hello, World!"를 응답으로 보내는 코드가 제거되었습니다.
2. getRequestKey 메소드가 추가되었습니다. 이 메소드는 HttpExchange 객체를 인자로 받아, 요청 메소드와 요청 경로를 결합하여 요청 키를 생성합니다.
3. handle 메소드에서 getRequestKey 메소드를 호출하여 요청 키를 생성하도록 변경되었습니다.

jihwooon committed:
HTTP 응답 처리 방식 개선

RequestHandler 클래스의 HTTP 응답 처리 방식이 변경되었습니다. 이전에는 요청 키만을 반환하였지만, 이제는 "Hello World"라는 메시지를 반환합니다.

주요 변경사항:
1. handle 메소드에서 "Hello World" 문자열을 생성하고, 이를 sendResponse 메소드를 통해 클라이언트에게 반환합니다.
2. sendResponse 메소드가 추가되었습니다. 이 메소드는 HTTP 응답 헤더를 설정하고, 응답 본문을 작성하여 클라이언트에게 반환합니다.

jihwooon committed:
IOException 처리 추가

RequestHandler 클래스에서 발생하는 IOException을 처리하는 코드를 추가했습니다. 이로써, 예외 발생 시 적절한 처리가 가능해집니다.

주요 변경사항:
1. RequestHandler 클래스의 try-catch 블록에 IOException 처리 코드 추가

jihwooon committed:
사칙연산 API 스펙 업데이트

OpenAPI 문서에서 사칙연산 API의 스펙을 업데이트하였습니다. 주로 입력 형식과 에러 메시지에 대한 설명을 보강하였습니다.

주요 변경사항:
1. 서버 URL의 들여쓰기를 수정하였습니다.
2. 사칙연산의 설명을 보강하고, 각 연산에 대한 설명을 추가하였습니다.
3. 400 에러 메시지의 설명을 "잘못된 입력 형식"으로 변경하고, 에러 메시지의 예시를 추가하였습니다.

jihwooon committed:
RequestHandler 클래스의 요청 처리 로직 변경

RequestHandler 클래스의 요청 처리 로직이 변경되었습니다. 이제 요청의 메소드와 경로에 따라 다른 응답을 반환합니다.

주요 변경사항:
1. `getRequestKey` 메소드가 `getRequestContent` 메소드로 변경되었으며, 이제 요청 본문을 읽어 반환합니다.
2. 새로운 `getRequestKey` 메소드가 추가되었으며, 이 메소드는 요청의 메소드와 경로를 확인하여 적절한 응답을 설정합니다.
3. `sendResponse` 메소드가 non-static으로 변경되었습니다.

jihwooon committed:
요청 처리 메소드 이름 변경 및 오류 메시지 추가

이 커밋에서는 요청을 처리하는 메소드의 이름을 변경하고, 응답 전송 중 발생하는 오류를 처리하는 코드를 추가했습니다.

주요 변경사항:
1. `getRequestKey` 메소드의 이름을 `processRequest`로 변경했습니다. 이는 메소드의 실제 기능을 더 잘 반영하도록 하기 위함입니다.
2. 응답 전송 중 오류가 발생하면, 이를 콘솔에 출력하는 코드를 추가했습니다. 이를 통해 오류 발생 시 더욱 명확한 디버깅이 가능해집니다.

ahastudio commented on app/src/main/java/com/example/demo/RequestHandler.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    private String getRequestContent(String requestContent, String method, String path) {
+        if (method.equals("POST") && path.equals("/calculations")) {
+            requestContent = "success";
+        } else if (method.equals("GET") && path.equals("/calculations")) {
+            requestContent = "success";
+        }
+
+        return requestContent;
```
if-else를 쓰고 있다면 어디선가 불필요한 일이 벌어지고 있는 겁니다. 이 코드는 몇 가지 이슈가 있습니다.

1. 매개변수의 값을 고치고 있습니다. 이는 반드시 피해야 하는 일입니다.
2. 매개변수가 아니라 임시 변수를 사용한다면, 불필요한 임시 변수입니다.
3. 어느 조건에도 맞지 않는다면 입력 받은 값을 그대로 다시 리턴합니다. 잠정적으로 버그일 가능성이 큽니다.

단순히 이런 코드가 돼야 합니다.

```suggestion
        if (method.equals("POST") && path.equals("/calculations")) {
            return "success";
        }
        if (method.equals("GET") && path.equals("/calculations")) {
            return "success";
        }

        return "기본 값 내지는 에러";
```

jihwooon replied:
if문 보다는 다형성으로 처리하면 좋을 것 같은데 객체를 어떻게 만들어 구분해야 할 지가 궁금합니다.

ahastudio replied:
다형성을 다루려면 먼저 동일한 형태의 반복으로 바꿔주는 게 좋습니다. 유사한 모양의 코드 반복 + 일치하지 않을 때의 기본 값 내지는 에러 정의가 먼저 선행돼야 하죠.

ahastudio commented on app/src/main/java/com/example/demo/RequestHandler.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    private String getRequestContent(HttpExchange exchange) throws IOException {
```
같은 이름의 메서드가 너무 멀리 떨어져 있습니다. 이렇게 위치가 멀면 오버로딩을 한다는 걸 알기 어렵습니다.

그리고 동일한 이름을 사용하면 안 되는 케이스입니다. 두 메서드는 왜 같은 이름을 쓰는지 알 수 없을 정도로 완전히 다릅니다.

jihwooon replied:
같은 이름의 메서드를 사용하고 있다는 것을 인지 하지 못 하고 있었습니다.
메서드 이름의 목적에 따라 이름을 구체화 해야겠습니다.

ahastudio replied:
메서드 순서를 항상 정리하세요. 아마 그냥 IntelliJ IDEA가 만들어준 메서드 순서를 그대로 사용하고 있을 거예요. 사용 순서대로 맞추면 좀 더 파악하기 쉽습니다. 순서에 대한 건 <Tidy First?>와 <클린 코드> 등에서 설명하고 있습니다.

jihwooon committed:
메소드와 경로 설정 로직 개선

RequestHandler 클래스에서 메소드와 경로를 설정하는 로직이 개선되었습니다. 이전에는 하나의 if-else 구문으로 처리되었지만, 이제는 각각의 if 구문으로 분리되어 처리됩니다.

주요 변경사항:
1. getRequestContent 메소드의 이름이 setRequestMethodAndPath로 변경되었습니다. 이는 메소드의 실제 기능을 더 잘 반영합니다.
2. POST와 GET 메소드에 대한 경로 설정 로직이 각각의 if 문으로 분리되었습니다. 이로 인해 코드의 가독성이 향상되었습니다.

jihwooon committed:
GET 메소드와 root 경로에 대한 응답 추가

RequestHandler 클래스에서 GET 메소드와 root 경로("/")에 대한 응답을 설정하였습니다. 이를 통해 해당 경로로의 요청에 대해 "hello world"라는 메시지를 반환하게 됩니다.

주요 변경사항:
1. RequestHandler 클래스의 setRequestMethodAndPath 메소드에 조건문 추가: 이제 메소드가 "GET"이고 경로가 "/"인 경우, 요청 내용을 "hello world"로 설정합니다.

ahastudio commented on app/src/main/java/com/example/demo/RequestHandler.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private String setRequestMethodAndPath(String requestContent, String method, String path) {
+        if (method.equals("GET") && path.equals("/")) {
+            requestContent = "hello world";
+        }
+        if (method.equals("POST") && path.equals("/calculations")) {
+            requestContent = "success";
+        }
+        if (method.equals("GET") && path.equals("/calculations")) {
+            requestContent = "success";
+        }
+
+        return requestContent;
```
이렇게 되면 `if` 검사를 항상 모조리 거쳐야 하기 때문에 비효율적입니다. `if` 블럭 안에서 바로 리턴하세요.

`requestContent` 매개변수를 고치는 것도 반드시 피해야 합니다. 매개변수는 항상 `final`이라고 생각하세요.

jihwooon committed:
요청 처리 로직 리팩토링 및 클래스 재구성

이 커밋에서는 HTTP 요청 처리 로직을 리팩토링하고 관련 클래스를 재구성하였습니다. 이를 통해 코드의 가독성과 유지보수성이 향상되었습니다.

주요 변경사항:
1. `RequestHandler` 클래스가 `presentation` 패키지로 이동되었습니다.
2. `RequestMethodHandler` 인터페이스가 새로 생성되었으며, `GetHomeRequest`와 `PostCalculationRequest` 클래스가 이를 구현하였습니다.
3. `RequestHandler`의 `setRequestMethodAndPath` 메소드 로직이 변경되었습니다. 이제 "Not Found Exception"을 반환하며, 각 요청 처리 로직은 새로 생성된 `GetHomeRequest`와 `PostCalculationRequest` 클래스에서 처리합니다.

jihwooon commented on docs/opeanapi.yaml:
```diff
@@ -0,0 +1,83 @@
+                                        type: number
+                                        description: 계산 결과값
+                                        example: 3
+                400:
+                    description: 잘못된 입력 형식
+                    content:
+                        application/json:
+                            schema:
+                                type: object
+                                properties:
+                                    error:
+                                        type: string
+                                        example: "사칙 연산이 올바르지 않습니다."
+        get:
+            responses:
+                200:
+                    description: 연산 계산 모음집
+                    content:
+                        application/json:
+                            schema:
+                                type: object
+                                properties:
+                                    calculations:
+                                        type: array
+                                        items:
+                                            type: object
+                                            properties:
+                                                number1:
+                                                    type: number
+                                                    example: 1
+                                                number2:
+                                                    type: number
+                                                    example: 2
+                                                operations:
+                                                    type: string
+                                                    example: "+"
+                                                result:
+                                                    type: number
+                                                    example: 3
+
```
api 문서는 그만!

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestMethodHandler.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.demo.presentation;
+
+public interface RequestMethodHandler {
+    String handler(String content);
+}
```
실제로 호출하려면 어떻게 작성해야 하는지 알려주세요

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,62 @@
+            return "success";
+        }
+
+        return "Not Fount Exception";
```
맞춤법 지적 감사합니다~!

jihwooon committed:
RequestHandler 메서드명 및 파라미터 정리

RequestHandler 클래스에서 메서드명과 파라미터를 명확하게 변경하였습니다.

주요 변경사항:
1. processRequest 메서드를 getProcessRequest로 변경하였습니다. 이는 메서드의 역할을 더욱 명확하게 표현하기 위함입니다.
2. sendResponse 메서드의 파라미터를 변경하였습니다. 이전에는 requestContent를 직접 전달하였지만, 이제는 processRequest의 결과를 전달합니다.
3. setRequestMethodAndPath와 sendResponse 메서드의 파라미터를 여러 줄로 나누어 가독성을 향상시켰습니다.

jihwooon committed:
POST /calculations 엔드포인트 로직 변경 및 GET /calculations 엔드포인트 수정

POST /calculations 엔드포인트의 로직이 변경되었으며, GET /calculations 엔드포인트의 반환 값을 수정하였습니다.

주요 변경사항:
1. POST /calculations 엔드포인트는 이제 두 개의 숫자를 받아 합계를 반환합니다.
2. GET /calculations 엔드포인트는 이제 requestContent를 반환합니다.

jihwooon committed:
요청 처리 메소드 및 변수명 개선

이 커밋에서는 요청을 처리하는 메소드와 변수의 이름을 더 명확하게 변경하였습니다. 이를 통해 코드의 가독성이 향상되었습니다.

주요 변경사항:
1. `getProcessRequest` 메소드의 이름을 `processRequest`로 변경하였습니다. 이는 메소드의 실제 기능을 더 잘 반영합니다.
2. `processRequest` 변수의 이름을 `responseContent`로 변경하였습니다. 이는 변수가 실제로 담고 있는 내용을 더 잘 설명합니다.
3. `processRequest` 메소드 내에서 두 숫자를 더하는 로직에 공백 라인을 추가하여 가독성을 향상시켰습니다.

jihwooon committed:
메소드와 경로 설정 함수의 매개변수 정리

RequestHandler 클래스의 setRequestMethodAndPath 함수에서 매개변수를 정리하였습니다.

주요 변경사항:
1. requestContent, method, path 매개변수를 각각 다른 줄로 분리하여 가독성을 향상시켰습니다.

jihwooon committed:
요청 처리 메소드와 경로 설정 로직 개선

RequestHandler 클래스에서 요청 처리 메소드와 경로 설정 로직이 개선되었습니다. 이전에는 요청 내용을 설정하는 로직이었지만, 이제는 응답 내용을 결정하는 로직으로 변경되었습니다.

주요 변경사항:
1. 'setRequestMethodAndPath' 메소드가 'determineResponseContent'로 이름이 변경되었습니다.
2. 이 메소드는 이제 요청 내용을 설정하는 것이 아니라 응답 내용을 결정하는 로직을 수행합니다.
3. 이에 따라 메소드의 반환 값도 요청 내용에서 응답 내용으로 변경되었습니다.

jihwooon committed:
오타 수정: "Not Found Exception" 메시지 반환

RequestHandler 클래스에서 발생하는 예외 메시지의 오타를 수정하였습니다.

주요 변경사항:
1. "Not Fount Exception"이라는 오타가 있던 메시지를 "Not Found Exception"으로 수정하였습니다.

jihwooon committed:
요청 처리 로직을 개선하고 코드 재구성

이번 커밋에서는 요청 처리 로직을 개선하고, 코드를 재구성하여 가독성을 향상시켰습니다.

주요 변경사항:
1. `App.java`에 불필요한 공백을 제거하였습니다.
2. `PostCalculationRequest.java`에서 요청 내용을 처리하는 로직을 개선하였습니다. 이제 숫자를 분리하고 더하는 로직이 이 클래스에 포함되어 있습니다.
3. `RequestHandler.java`에서 요청 처리 로직을 개선하였습니다. 이제 각 요청 경로에 대한 처리 로직이 각각의 클래스(`GetHomeRequest`, `PostCalculationRequest`)로 분리되어 있습니다.
4. `RequestMethodHandler.java`에 불필요한 공백을 제거하였습니다.

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        return "Not Found Exception";
+    }
```
Map을 사용하는 방법도 있지만 다른 방법으로 사용 할 수 있나요?

jihwooon committed:
GET 요청 처리 로직 개선

GET 요청에 대한 처리 로직을 개선하였습니다. 이제 "/calculations" 경로에 대한 GET 요청이 들어올 경우, 새로운 ListCalculationRequest 클래스의 handler 메서드가 호출됩니다.

주요 변경사항:
1. ListCalculationRequest 클래스 추가: RequestMethodHandler 인터페이스를 구현하며, handler 메서드에서는 "list" 문자열을 반환합니다.
2. RequestHandler 클래스 수정: "/calculations" 경로에 대한 GET 요청 처리 로직을 변경하였습니다. 이제 ListCalculationRequest의 handler 메서드가 호출되어 결과를 반환합니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,74 @@
+
+        try (OutputStream outputStream = exchange.getResponseBody()) {
+            outputStream.write(bytes);
+        } catch (IOException e) {
+            System.err.println("응답 전송 중 오류 발생: " + e.getMessage());
+            throw e;
+        }
```
대부분은 내부에서 try-catch를 쓰지 말고, 가장 밖에서 예외를 잡아서 사용자가 뭔가 대응할 수 있게 합니다. 사용자가 대응할 수 없는 예외는 쓸모가 없습니다.

개발할 때 사용하는 로깅 용도에 가깝다면, Sentry 등 관련 솔루션을 활용하는 게 훨씬 좋습니다.

정말 저수준에서 뭔가를 하고 싶다면 그대로 throw하지 말고 Chained Exception을 사용하세요. https://docs.oracle.com/javase/tutorial/essential/exceptions/chained.html

ahastudio replied:
https://github.com/dal-lab/wild-backend-calculator/pull/1#discussion_r1784168834

좀 이상하다 싶으면 이런 식으로 강력 대응해야 합니다. 😆

jihwooon replied:
try-catch 문을 언제 사용해야 하는지에 대해서 궁금합니다.
여러 코드를 접하다보면 try-catch문을 종종 사용하지만 정말로 이게 왜 써야 하는지에 대해서는 명쾌한 답을 얻지 못 했습니다.

ahastudio replied:
기본은 이겁니다: 예외는 따로 catch하지 않습니다.

다음은 이겁니다: 사용자가 대응할 수 있는 곳에서 catch합니다. 예를 들어, HTTP 서버라면 UI Layer에서 catch해서 적절한 HTTP Status Code로 돌려줍니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,74 @@
+        return "Not Found Exception";
+    }
+
+    private void sendResponse(HttpExchange exchange, String responseContent)
+            throws IOException {
+        byte[] bytes = responseContent.getBytes();
+
+        exchange.sendResponseHeaders(200, bytes.length);
+
+        try (OutputStream outputStream = exchange.getResponseBody()) {
+            outputStream.write(bytes);
+        } catch (IOException e) {
+            System.err.println("응답 전송 중 오류 발생: " + e.getMessage());
+            throw e;
+        }
+    }
```
@coderabbitai // try-catch에서 예외를 println하고 그 예외를 그대로 다시 던지는 건 bad practice입니다. 이런 코드를 추천한 이유가 있나요? 스스로 생각하기에도 나쁘다고 생각하면 이렇게 하지 마세요.

jihwooon replied:
예외 발생 시 예외를 처리하지 않고 호출한 메서드로 상위 레벨에서 처리 할 수 있다는 말이 무슨 말인지 설명해주시면 감사하겠습니다.

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    private String getRequestContent(HttpExchange exchange) throws IOException {
+        InputStream inputStream = exchange.getRequestBody();
+        return new String(inputStream.readAllBytes());
+    }
```
성능에 대해서는 나중에 물어볼 때 대답해줬으면 좋겠어

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,74 @@
+import java.io.OutputStream;
+import java.net.URI;
+
+public class RequestHandler implements HttpHandler {
```
클래스 상속 하지 않기 위해서 `final`을 붙여야 하군요~!

jihwooon committed:
Chained Exceptions를 적용한 ResponseWriteException 클래스 추가 및 적용

응답 전송 중 발생하는 IOException을 처리하기 위해 새로운 예외 클래스를 추가하고, 이를 적용하였습니다.

주요 변경사항:
1. ResponseWriteException 클래스 추가: IOException을 처리하기 위한 새로운 RuntimeException 서브 클래스입니다.
2. RequestHandler 클래스에 ResponseWriteException 적용: 응답 전송 중 IOException이 발생하면 ResponseWriteException을 던집니다.
3. HTTP 상태 코드를 상수로 변경: Magic number였던 200을 HTTP_OK라는 상수로 변경하였습니다.

jihwooon committed:
NotFoundException 추가 및 RequestHandler 수정

NotFoundException 클래스를 추가하고, RequestHandler에서 "Not Found Exception" 문자열을 반환하는 대신 NotFoundException을 던지도록 수정하였습니다.

주요 변경사항:
1. NotFoundException 클래스 추가: 요청한 리소스를 찾을 수 없을 때 발생하는 예외를 처리하기 위한 클래스입니다.
2. RequestHandler 수정: 요청한 리소스를 찾을 수 없을 때 "Not Found Exception" 문자열을 반환하는 대신 NotFoundException을 던집니다.

jihwooon commented on app/src/main/java/com/example/demo/exception/NotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+
+public class NotFoundException extends RuntimeException {
+
+    public NotFoundException(String message) {
+        super(message);
+    }
```
기본 생성자에 메시지를 작성하는 것과 message를 인자로 받아서 message를 명시적으로 나타내는 것 중 어떤 것을 권장하나요?

jihwooon committed:
예외 클래스에 대한 주석 추가

이 커밋에서는 두 개의 예외 클래스에 대한 주석을 추가했습니다. 이 주석들은 클래스와 생성자의 기능을 명확하게 설명하고 있습니다.

주요 변경사항:
1. NotFoundException 클래스에 대한 주석 추가: 이 클래스는 찾을 수 없는 요청이 발생했을 때 던져지는 예외를 나타냅니다.
2. ResponseWriteException 클래스에 대한 주석 추가: 이 클래스는 응답 작성 중 발생한 예외를 나타냅니다.

jihwooon committed:
요청 핸들러의 리팩토링

RequestHandler 클래스에서 요청 처리 방식을 개선하였습니다. 이전에는 각 요청 방식과 경로에 따라 분기 처리를 하였지만, 이를 Map을 이용하여 간결하게 처리하도록 변경하였습니다.

주요 변경사항:
1. 요청 방식과 경로를 키로 가지는 Map 생성: 각 요청 방식과 경로에 따른 핸들러를 Map에 저장하였습니다.
2. 요청 처리 방식 변경: 요청 방식과 경로를 키로 가지는 Map에서 핸들러를 찾아 처리하도록 변경하였습니다.

jihwooon committed:
요청 처리 로직 리팩토링

RequestHandler 클래스의 요청 처리 로직이 리팩토링 되었습니다. 이제 요청 키를 생성하는 로직과 요청을 처리하는 로직이 분리되어 코드의 가독성이 향상되었습니다.

주요 변경사항:
1. processRequest 메소드가 getRequestKey와 getRequestHandler 두 메소드로 분리되었습니다.
2. getRequestKey 메소드는 HttpExchange 객체로부터 요청 메소드와 경로를 추출하여 요청 키를 생성합니다.
3. getRequestHandler 메소드는 요청 키를 이용하여 적절한 요청 처리 핸들러를 찾아 요청을 처리합니다.

jihwooon commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,80 @@
+            throws IOException {
+        byte[] bytes = responseContent.getBytes();
+
+        exchange.sendResponseHeaders(HTTP_OK, bytes.length);
+
+        try {
+            OutputStream outputStream = exchange.getResponseBody();
+            outputStream.write(bytes);
+        } catch (IOException e) {
+            String message = e.getMessage();
+            throw new ResponseWriteException(message);
+        }
+    }
```
@coderabbitai Oracle 공식 문서 중 Chained Exceptions 바탕으로 예외 케이스를 던지는 것을 적용 했습니다. 추천 한 코드는 로깅처리가 되지 않음으로 삼가해주시길 바랍니다.

jihwooon committed:
계산기 기능에 사칙연산 추가

PostCalculationRequest 핸들러에서 단순 덧셈만 지원하던 기능을 사칙연산을 지원하도록 확장하였습니다.

주요 변경사항:
1. 입력 문자열을 파싱하여 연산자를 추출하도록 변경하였습니다.
2. 추출한 연산자에 따라 덧셈, 뺄셈, 곱셈, 나눗셈을 수행하도록 switch 문을 추가하였습니다.
3. 지원하지 않는 연산자가 입력되었을 경우 IllegalArgumentException을 발생시키도록 하였습니다.

jihwooon committed:
계산기 연산 로직을 별도의 클래스로 분리

기존에 PostCalculationRequest 클래스 내에 있던 계산 로직을 각각의 연산자에 대응하는 별도의 클래스로 분리하였습니다.

주요 변경사항:
1. OperatorPlus, OperatorMinus, OperatorMultiply, OperatorDivide 클래스 추가: 각 클래스는 해당 연산자에 대한 계산 로직을 담당합니다.
2. PostCalculationRequest 클래스 수정: 기존에 내장되어 있던 계산 로직을 새로 생성한 연산자 클래스의 메소드를 호출하는 방식으로 변경하였습니다.

jihwooon committed:
계산 로직을 Calculator 클래스로 이동

기존에 PostCalculationRequest 클래스 내에 있던 계산 로직을 별도의 Calculator 클래스로 이동하였습니다. 이로써 각 클래스의 역할이 명확해졌습니다.

주요 변경사항:
1. Calculator 클래스 생성: 이 클래스는 두 숫자와 연산자를 입력받아 계산 결과를 반환하는 calculate 메소드를 가지고 있습니다.
2. PostCalculationRequest 클래스 수정: 기존에 내장되어 있던 계산 로직을 Calculator 클래스의 calculate 메소드를 호출하는 것으로 변경하였습니다.

jihwooon committed:
계산기 함수의 반환 타입 변경

기존에 문자열로 반환하던 계산기 함수들이 정수형으로 반환하도록 수정하였습니다.

주요 변경사항:
1. Calculator 클래스의 calculate 메소드가 이제 정수를 반환합니다.
2. OperatorDivide, OperatorMinus, OperatorMultiply, OperatorPlus 클래스의 calculate 메소드들이 이제 정수를 반환합니다.
3. PostCalculationRequest 클래스에서 calculate 메소드의 결과를 문자열로 변환하여 반환합니다.

jihwooon committed:
계산기 코드 리팩토링 및 Operator 인터페이스 도입

기존의 계산기 코드를 리팩토링하고, 연산자 클래스들이 Operator 인터페이스를 구현하도록 변경하였습니다.

주요 변경사항:
1. Calculator 클래스에서 연산자를 Map으로 관리하도록 변경하였습니다. 이를 통해 연산자를 동적으로 관리할 수 있게 되었습니다.
2. 각 연산자 클래스(OperatorPlus, OperatorMinus, OperatorMultiply, OperatorDivide)가 Operator 인터페이스를 구현하도록 변경하였습니다. 이를 통해 각 연산자 클래스의 일관성을 보장하였습니다.
3. 각 연산자 클래스의 calculate 메소드가 static에서 non-static으로 변경되었습니다. 이를 통해 각 연산자 클래스의 인스턴스를 생성하여 사용할 수 있게 되었습니다.

jihwooon committed:
계산기 예외 처리 로직 추가

계산기 애플리케이션에 예외 처리 로직을 추가하였습니다. 이를 통해 유효하지 않은 연산자나 0으로 나누는 경우에 대한 예외를 처리할 수 있게 되었습니다.

주요 변경사항:
1. calculate 메소드에 유효하지 않은 연산자에 대한 예외 처리 로직 추가: 연산자가 null이거나 operators에 없는 경우 IllegalArgumentException을 발생시킵니다.
2. calculate 메소드에 0으로 나누는 경우에 대한 예외 처리 로직 추가: 연산자가 나누기이고 number2가 0인 경우 ArithmeticException을 발생시킵니다.
3. ArithmeticException 클래스 추가: RuntimeException을 상속받는 새로운 예외 클래스를 추가하였습니다. 이 예외는 0으로 나누는 경우에 발생합니다.

jihwooon committed:
나누기 연산 예외 처리 클래스 변경

Calculator 클래스에서 발생하는 나누기 연산 예외를 처리하는 클래스를 변경하였습니다. 기존 ArithmeticException에서 CalculatorArithmeticException으로 변경하였습니다.

주요 변경사항:
1. Calculator 클래스에서 나누기 연산 예외를 처리하는 부분의 예외 클래스를 CalculatorArithmeticException으로 변경
2. ArithmeticException 클래스 삭제
3. CalculatorArithmeticException 클래스 추가

jihwooon committed:
Jackson 라이브러리 추가 및 Calculation 클래스 생성

이 커밋에서는 Jackson 라이브러리를 추가하고, 계산 결과를 저장하는 Calculation 클래스를 생성했습니다.

주요 변경사항:
1. build.gradle.kts 파일에 Jackson 라이브러리를 추가했습니다. 이 라이브러리는 JSON 데이터를 Java 객체로 변환하는데 사용됩니다.
2. Calculation 클래스를 생성했습니다. 이 클래스는 두 개의 숫자, 연산자, 그리고 그 결과를 저장하는 필드를 가지고 있습니다.

jihwooon committed:
계산기 애플리케이션에 계산 결과 저장 기능 추가

계산기 애플리케이션에 계산 결과를 저장하고 반환하는 기능을 추가하였습니다. 이를 위해 Calculation과 CalculationRepository 클래스를 새로 생성하였고, 기존의 코드를 수정하였습니다.

주요 변경사항:
1. Calculation 클래스 생성: 계산에 사용된 숫자, 연산자, 그리고 결과를 저장하는 클래스입니다.
2. CalculationRepository 클래스 생성: 계산 결과를 저장하는 저장소 역할을 하는 클래스입니다.
3. Calculator 클래스 수정: 계산 결과를 Calculation 객체로 반환하도록 수정하였습니다. 또한, 계산 결과를 CalculationRepository에 저장하는 로직을 추가하였습니다.
4. PostCalculationRequest 클래스 수정: 요청 본문을 CalculationRequestDto로 변환하고, 응답을 CalculationResponseDto로 반환하도록 수정하였습니다.
5. CalculationRequestDto, CalculationResponseDto 클래스 생성: 요청과 응답을 처리하기 위한 DTO 클래스를 생성하였습니다.

jihwooon committed:
계산기 애플리케이션에 계산 이력 조회 기능 추가

계산기 애플리케이션에 사용자가 이전에 수행한 계산의 이력을 조회할 수 있는 기능을 추가하였습니다.

주요 변경사항:
1. Calculator 클래스에 getCalculationList 메소드 추가: 이 메소드는 CalculationRepository에서 모든 계산 이력을 조회합니다.
2. CalculationRepository 클래스에 findAll 메소드 추가: 이 메소드는 저장된 모든 계산 이력을 반환합니다.
3. ListCalculationRequest 클래스에 계산 이력을 JSON 형태로 변환하여 반환하는 로직 추가.
4. CalculationListRequestDto 클래스 추가: 이 클래스는 계산 이력을 담는 DTO입니다.

jihwooon committed:
계산기 애플리케이션의 반환 타입 및 DTO 변경

계산기 애플리케이션의 반환 타입을 변경하고, 이에 따른 DTO의 수정이 이루어졌습니다.

주요 변경사항:
1. Calculator 클래스의 calculate 메소드의 반환 타입이 Calculation에서 int로 변경되었습니다.
2. 이에 따라 PostCalculationRequest 클래스에서 반환하는 CalculationResponseDto의 생성자 인자가 수정되었습니다.
3. CalculationResponseDto와 CalculationListRequestDto에서 불필요한 필드를 제거하고, CalculationListResponseDto를 새로 생성하여 필요한 필드만을 가지도록 하였습니다.
4. CalculationRepository의 findAll 메소드에서 calculations 리스트를 출력하는 로그를 추가하였습니다.

jihwooon committed:
DTO 클래스 변경 및 요청/응답 핸들러 수정

이 커밋에서는 DTO 클래스의 구조를 변경하고, 이에 따라 요청/응답 핸들러를 수정하였습니다.

주요 변경사항:
1. CalculationListRequestDto와 CalculationListResponseDto의 역할이 바뀌었습니다. CalculationListRequestDto는 이제 개별 계산 결과를 나타내는 DTO로 사용되며, CalculationListResponseDto는 이제 계산 결과 리스트를 나타내는 DTO로 사용됩니다.
2. ListCalculationRequest와 PostCalculationRequest 클래스에서 CalculationListRequestDto 대신 CalculationListResponseDto를 사용하도록 수정하였습니다.
3. CalculationListRequestDto와 CalculationListResponseDto의 필드와 메소드가 변경되었습니다. CalculationListRequestDto는 이제 개별 계산 결과를 나타내는 필드를 가지며, CalculationListResponseDto는 계산 결과 리스트를 나타내는 필드를 가집니다.

jihwooon commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,49 @@
+
+    private final Map<String, Operator> operators = new HashMap<>();
+
+    public Calculator() {
+        operators.put("+", new OperatorPlus());
+        operators.put("-", new OperatorMinus());
+        operators.put("*", new OperatorMultiply());
+        operators.put("/", new OperatorDivide());
+    }
```
팩토리 패턴이 어떻게 createoperators를 메서드 형태로 만들어 반환하는지 알려줘

jihwooon committed:
계산기 코드 리팩토링 및 계산 저장 로직 분리

계산기의 주요 기능인 계산 수행 및 계산 결과 저장 로직이 분리되었습니다. 이를 통해 코드의 가독성과 유지보수성이 향상되었습니다.

주요 변경사항:
1. `calculate` 메소드 내부에서 수행하던 연산자 유효성 검사를 `validateOperator` 메소드로 분리하였습니다.
2. 계산 수행 로직을 `performCalculation` 메소드로 분리하였습니다.
3. 계산 결과 저장 로직을 `saveCalculation` 메소드로 분리하였습니다.

jihwooon commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,49 @@
+        calculationRepository.add(calculation);
+
+        return calculation.getResult();
+    }
```
private 메서드를 만드는 순간 srp 원칙을 어긋 나는데 이 방법이 올바른지 확인해주세요

jihwooon replied:
private 메서드는 테스트 코드를 작성 시 테스트가 어려워지는 경우가 있습니다. 이럴 때 따로 클래스를 분리해서 해당 메서드를 불러오는 게 좋을 것 같다는 생각 듭니다.

jihwooon committed:
CalculationRepository를 싱글톤 패턴으로 변경

CalculationRepository의 인스턴스 생성 방식을 변경하였습니다. 이제 CalculationRepository는 싱글톤 패턴을 따르게 됩니다.

주요 변경사항:
1. CalculationRepository의 생성자를 private으로 변경하여 외부에서의 직접적인 인스턴스 생성을 막았습니다.
2. getInstance 메소드를 추가하여, 이 메소드를 통해 CalculationRepository의 인스턴스를 얻을 수 있게 하였습니다.
3. Calculator 클래스에서 CalculationRepository의 인스턴스를 얻는 방식을 getInstance 메소드를 사용하도록 변경하였습니다.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+        Calculation calculation = saveCalculation(number1, number2,
+                operatorSymbol, result);
+
+        return calculation.getResult();
```
특별한 이유가 없다면 그냥 Calculation 객체를 리턴하시면 UI Layer에서 값을 만들 때 유연하게 활용할 수 있습니다.

jihwooon replied:
`application Layer`에서 미리 사용자에게 전달 할 내용을 바로 리턴하기보다는 `UI Layer` 에서 값을 전달 할 수 있게 하는게 더 좋은 의도 인지가 궁금합니다.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+        int result = performCalculation(number1, number2, operatorSymbol);
+
+        Calculation calculation = saveCalculation(number1, number2,
+                operatorSymbol, result);
```
`performCalculation`이나 `saveCalculation`는 표현이 모호하면서 큰 가치를 갖지 않습니다. 오히려 중요한 문제를 숨기고 있죠.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+            String operatorSymbol) {
+        Operator operator = operators.get(operatorSymbol);
+
+        if (operator instanceof OperatorDivide && number2 == 0) {
+            throw new CalculatorArithmeticException("나누기는 0으로 나눌 수 없습니다.");
+        }
```
이렇게 하면 Operator를 쓰는 이유가 없습니다. 타입을 알아내서 뭔가 처리하는 건 객체지향에 반하는 결정입니다.

ahastudio commented on app/src/main/java/com/example/demo/application/OperatorDivide.java:
```diff
@@ -0,0 +1,9 @@
+public class OperatorDivide implements Operator {
+
+    @Override
+    public int calculate(int number1, int number2) {
+        return number1 / number2;
+    }
```
바로 여기서 number2가 0인지 검사해야 합니다. 여기서 예외를 던져야 하고, 이 예외는 UI Layer까지 전달될 겁니다.

ahastudio commented on app/src/main/java/com/example/demo/exception/ResponseWriteException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * 응답 작성 중 발생한 예외를 나타냅니다.
+ */
+public class ResponseWriteException extends RuntimeException {
```
이 예외는 의미가 없습니다. 우리가 사용자에게 응답할 수 없다면 이 예외는 영원히 쓰일 수 없습니다.

정말 정말 꼭 이걸 쓰고 싶다면 반드시 Chained Exception이어야 하고, 다시 반복하면, 가능하면 아예 쓰지 않는 것을 권합니다.

ahastudio commented on app/src/main/java/com/example/demo/exception/ResponseWriteException.java:
```diff
@@ -0,0 +1,16 @@
+     *
+     * @param message 상세 메시지
+     */
+    public ResponseWriteException(String message) {
```
우리가 커스텀 예외를 만들면서 가장 먼저 고려해야 하는 건 message가 아니라 cause입니다. cause가 있어야 다시 예외를 계속 추적할 수 있습니다. 예외가 발생했는데, 이런 예외로 다시 바꿔서 보내면 어디서 왜 예외가 발생했는지 알 길이 없어집니다.

문제 상황은 Sentry나 여타 다른 도구를 사용해서 지옥 끝까지 추적해야 하고, 이런 예외가 정보를 지워버리면 우리는 재현하기 어려운 문제 때문에 고통받기 쉽습니다.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+        return calculation.getResult();
+    }
+
+    private Calculation saveCalculation(int number1, int number2,
+            String operatorSymbol, int result) {
+        Calculation calculation = new Calculation(number1, number2,
+                operatorSymbol,
+                result);
+
+        calculationRepository.add(calculation);
+        return calculation;
+    }
+
+    private int performCalculation(int number1, int number2,
+            String operatorSymbol) {
+        Operator operator = operators.get(operatorSymbol);
+
+        if (operator instanceof OperatorDivide && number2 == 0) {
+            throw new CalculatorArithmeticException("나누기는 0으로 나눌 수 없습니다.");
+        }
+        int result = operator.calculate(number1, number2);
+        return result;
+    }
+
+    private void validateOperator(String operatorSymbol) {
+        if (operatorSymbol == null || !operators.containsKey(operatorSymbol)) {
+            throw new IllegalArgumentException(
+                    "Invalid operator: " + operatorSymbol);
+        }
+    }
```
메서드 순서를 맞춰주세요. 이렇게 역순으로 되어 있으면 코드를 읽기 어렵습니다.

`calculate` 메서드에서 사용하는 순서는 `validateOperator` -> `performCalculation` -> `saveCalculation` 입니다.

jihwooon replied:
인텔리제이에서 정해 준 순서로만 만들었는데 아직 이 부분이 습관화 되지 않아서 계속해서 순서를 고려 하지 않았던 것 같습니다.
코드 읽기 사람이 읽기 쉽게 정렬하는 습관도 이번 기회에 가져가야 할 부분인 것 같습니다.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+        return calculation;
+    }
+
+    private int performCalculation(int number1, int number2,
```
이미 `calculate`라는 이름의 메서드가 있는데, 그 안에서 `perform`이란 표현을 쓰는 건 둘 사이에 어떤 관계가 있는 건지 고심해야 합니다. 굳이 그런 관계를 만들 필요가 없습니다.

if 문이 잘못 들어온 걸 제외하고 나면, 이 메서드는 한 줄로도 표현 가능할 정도로 단순합니다. 메서드가 이해를 돕는지, 오히려 저해하는지 따져보면 좋습니다. 리팩터링에 inline이 있는 이유기도 하죠.

ahastudio commented on app/src/main/java/com/example/demo/application/Calculator.java:
```diff
@@ -0,0 +1,66 @@
+        return calculation.getResult();
+    }
+
+    private Calculation saveCalculation(int number1, int number2,
```
도메인 객체를 만든 상태에서 바로 이렇게 메서드로 분할하면 도메인 객체를 잘 활용하기 어려워집니다. 예를 들어, 도메인 객체를 적극 활용하기로 하면 이런 것도 가능합니다.

```java
Calculation calculation = Calculation.of(operator);
calculation.add(number1);
calculation.add(number2);
calculation.calculate();
calculationRepository.add(calculation);
```

체이닝을 선호한다면 이런 모양도 가능하죠.

```java
Calculation calculation = Calculation.of(operator).add(number1).add(number2).end();
calculationRepository.add(calculation);
```

설계를 유연하게 만드는데 도움이 되는 경계를 찾는 게 중요합니다. 그래야 강의에서 다룬 것 이상으로 더 발전할 수 있습니다. 너무 빨리 나누려고 하지 마세요. 진짜로 의미 있는 걸 발견했을 때 옮겨보세요.

jihwooon replied:
도메인 객체를 잘 활용하려면 뭔가에 여기서 무언가의 패턴이 발견 되면 위에 주신 내용을 활용 할 수 있을 것 같습니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,76 @@
+            outputStream.write(bytes);
+        } catch (IOException e) {
+            String message = e.getMessage();
+            throw new ResponseWriteException(message);
```
이 예외는 쓸 곳이 없습니다.

반대의 강박을 가지고 작업을 해보세요: “catch를 하나도 쓸 수 없다면 어떻게 될까?”

jihwooon replied:
이 고민은 해봐야겠습니다. "나는 왜 `catch`를 고집해서 쓰는지를 스스로 생각해보고 이에 바탕으로 피드백 주신 방법으로 발전 시키도록 해보겠습니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/dto/CalculationListResponseDto.java:
```diff
@@ -0,0 +1,30 @@
+        return new CalculationListResponseDto(
+                calculations.stream()
+                        .map(calculation -> {
+                                    return new CalculationListRequestDto(
```
Response 안에 Request가 있을 수 없습니다. 그냥 `CalculationListItemResponseDto` 같은 이름이 필요할 겁니다.

jihwooon replied:
`CalculationListRequestDto`가 같은 컬럼 가지고 있어서 재활용으로 사용했는데 이는 좋은 방법이 아닌 것 같아 다음부터는 Response 에 대해 같은 이름을 짓는 걸로 해서 만들어야 겠습니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/dto/CalculationListResponseDto.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CalculationListResponseDto {
```
강의에선 Java 11을 써서 그냥 class로 했는데, Java 14부터는 record를 쓸 수 있습니다. record를 강력히 추천해요.

jihwooon replied:
기존 class와 recode 어떤 차이점이 있는지, 왜 recode가 좀 더 좋은지를 조사해보고 활용해보겠습니다.

jihwooon committed:
saveCalculation 메서드에 result 값을 리턴 대신 Calculation 객체를 리턴하도록 수정

jihwooon committed:
CalculationListRequestDto 클래스를 레코드로 변경

CalculationListRequestDto 클래스가 불변성을 가지고 있어, 레코드로 변경하였습니다. 이로써 코드의 가독성과 유지보수성이 향상되었습니다.

주요 변경사항:
1. CalculationListRequestDto 클래스를 레코드로 변경: 이전에는 각 필드에 대해 getter를 명시적으로 선언해야 했지만, 레코드를 사용함으로써 이를 생략할 수 있게 되었습니다.
2. 생성자와 getter 메소드 제거: 레코드를 사용하면 필드에 대한 생성자와 getter 메소드가 자동으로 생성되므로, 이를 직접 작성할 필요가 없어졌습니다.

jihwooon committed:
계산기 메소드 순서 재배치 및 리팩토링

Calculator 클래스의 메소드 순서를 재배치하고, 코드를 리팩토링하였습니다. 이를 통해 코드의 가독성과 유지보수성이 향상되었습니다.

주요 변경사항:
1. `validateOperator` 메소드와 `saveCalculation` 메소드의 위치를 서로 바꾸었습니다. 이로써 메소드 순서가 더욱 자연스러워졌습니다.
2. `saveCalculation` 메소드에서는 이제 Calculation 객체를 생성하고 저장하는 역할만 수행하며, 이 객체를 반환합니다.
3. `validateOperator` 메소드는 연산자 유효성 검사만 수행하며, 유효하지 않은 연산자에 대해서는 예외를 발생시킵니다.

jihwooon committed:
0으로 나누는 경우 예외 처리 위치 변경

기존에는 Calculator 클래스에서 0으로 나누는 경우를 체크하고 있었으나, 이를 OperatorDivide 클래스로 이동하였습니다.

주요 변경사항:
1. Calculator 클래스에서 0으로 나누는 경우의 예외 처리 코드를 제거하였습니다.
2. OperatorDivide 클래스의 calculate 메소드에서 0으로 나누는 경우 예외를 발생시키도록 변경하였습니다.

jihwooon committed:
계산기 코드 리팩토링

Calculator 클래스의 계산 및 저장 메소드를 간소화하였습니다.

주요 변경사항:
1. performCalculation 메소드와 saveCalculation 메소드를 제거하였습니다.
2. 이 두 메소드의 로직을 calculate 메소드에 통합하였습니다.
3. calculate 메소드에서 직접 Calculation 객체를 생성하고 저장하도록 변경하였습니다.

jihwooon committed:
계산 결과를 표현하는 DTO 클래스 수정 및 추가

계산 결과를 표현하는 DTO 클래스들이 수정되고 추가되었습니다. 이를 통해 계산 결과를 더욱 명확하게 표현할 수 있게 되었습니다.

주요 변경사항:
1. Calculation 클래스에 toString 메소드 추가: 계산 결과를 문자열로 표현할 수 있게 되었습니다.
2. ListCalculationRequest 클래스에서 jsonResponse 변수 추가: 직렬화된 JSON 응답을 명확하게 표현합니다.
3. CalculationListItemResponseDto 클래스 추가: 계산 결과 항목을 표현하는 새로운 DTO 클래스입니다.
4. CalculationListRequestDto 클래스 삭제: 이전에 사용하던 DTO 클래스를 삭제하였습니다.
5. CalculationListResponseDto 클래스 수정: CalculationListItemResponseDto 클래스를 사용하여 계산 결과 리스트를 표현합니다.

jihwooon committed:
Calculation 클래스에서 toString 메서드 제거

Calculation 클래스에서 toString 메서드를 제거하였습니다. 이 메서드는 객체의 상태를 문자열로 변환하는데 사용되었으나, 현재는 필요하지 않아 제거하였습니다.

주요 변경사항:
1. Calculation 클래스의 toString 메서드 제거: 이 메서드는 객체의 상태를 문자열로 변환하는 기능을 가졌으나, 현재는 사용되지 않아 제거하였습니다.

jihwooon committed:
CalculationListResponseDto 코드 리팩토링

CalculationListResponseDto 클래스에서 불필요한 메소드를 제거하고, 코드를 간결하게 수정하였습니다.

주요 변경사항:
1. 불필요한 getCalculations() 메소드를 제거하였습니다.
2. Collectors.toList()를 toList()로 변경하여 코드를 간결하게 만들었습니다.

jihwooon committed:
코드 간결화를 위한 리팩토링

이번 커밋에서는 코드의 간결화를 위해 불필요한 변수 할당과 예외 처리를 제거하였습니다.

주요 변경사항:
1. ListCalculationRequest 클래스에서 jsonResponse 변수 할당을 제거하고, 직접 return하도록 수정하였습니다.
2. RequestHandler 클래스에서 응답을 작성하는 과정에서 발생할 수 있는 IOException을 별도로 처리하지 않고, 메서드 시그니처에 throws로 선언하였습니다.

jihwooon committed:
응답 스트림 쓰기 과정 간소화

HttpExchange의 응답 바디에 바이트를 쓰는 과정에서 불필요한 변수 할당을 제거하였습니다. 이를 통해 코드의 가독성을 향상시켰습니다.

주요 변경사항:
1. OutputStream 객체 생성과 해당 객체를 통한 바이트 쓰기를 한 줄로 합쳤습니다.

jihwooon committed:
요청 핸들러의 예외 처리 로직 개선

RequestHandler 클래스에서 요청 처리 로직의 예외 처리 부분을 개선했습니다. 이전에는 요청 핸들러가 null인 경우에도 처리를 시도하였으나, 이제는 null 체크 후 적절한 예외를 발생시킵니다.

주요 변경사항:
1. 요청 핸들러가 null인 경우 즉시 NotFoundException을 발생시키도록 변경
2. 요청 핸들러가 존재하는 경우에만 요청 처리를 시도하도록 로직 수정

jihwooon committed:
CalculationResponseDto에서 getResult 메서드 제거

CalculationResponseDto에서 불필요한 getResult 메서드를 제거하였습니다.

주요 변경사항:
1. CalculationResponseDto의 getResult 메서드 제거: 이 메서드는 사용되지 않으므로 제거하였습니다.

jihwooon committed:
요청 핸들러의 키 값 리팩토링

RequestHandler 클래스에서 HTTP 요청을 처리하는 핸들러의 키 값을 상수로 리팩토링하였습니다.

주요 변경사항:
1. "GET /", "POST /calculations", "GET /calculations" 문자열을 각각 KEY, KEY1, KEY2 상수로 변경하였습니다.
2. handlers.put 메소드 호출 시, 직접 문자열을 입력하는 대신 상수를 사용하여 가독성을 향상시켰습니다.

jihwooon committed:
계산 결과를 반환하는 메소드 추가

CalculationResponseDto 클래스에 계산 결과를 반환하는 getResult 메소드를 추가했습니다. 이 메소드는 계산 결과를 외부로 전달하는 역할을 합니다.

주요 변경사항:
1. CalculationResponseDto 클래스에 getResult 메소드 추가: 이 메소드는 계산 결과를 반환합니다.

jihwooon committed:
나누기 연산 수정 및 불필요한 예외 제거

이 커밋에서는 나누기 연산의 조건을 수정하고, 더 이상 사용되지 않는 예외를 제거하였습니다.

주요 변경사항:
1. OperatorDivide 클래스에서 나누기 연산의 조건을 수정하였습니다. 이전에는 분모가 0일 때 예외를 발생시켰지만, 이제는 분자와 분모 모두 0일 때 예외를 발생시킵니다.
2. 더 이상 사용되지 않는 ResponseWriteException 클래스를 제거하였습니다.

jihwooon committed:
계산 객체 생성 방식 변경

Calculator 클래스에서 Calculation 객체를 생성하는 방식이 변경되었습니다. 이제 Calculation 클래스 내부에 static factory method를 사용하여 객체를 생성합니다.

주요 변경사항:
1. Calculator 클래스에서 Calculation 객체 생성 방식 변경: 직접 new 키워드를 사용하는 대신, Calculation 클래스의 createCalculation 메소드를 사용하여 객체를 생성합니다.
2. Calculation 클래스에 createCalculation 메소드 추가: 이 메소드는 Calculation 객체를 생성하고 반환합니다.

jihwooon committed:
계산 관련 클래스에 대한 수정 및 리팩토링

이번 커밋에서는 계산 관련 클래스들의 구조를 수정하고, 코드를 리팩토링하였습니다.

주요 변경사항:
1. Calculation 클래스에 toString 메소드 추가: Calculation 객체의 상태를 문자열로 표현할 수 있게 되었습니다.
2. CalculationListItemResponseDto 클래스를 record에서 일반 클래스로 변경: 이를 통해 더 유연한 코드 작성이 가능해졌습니다.
3. CalculationListResponseDto 클래스를 일반 클래스에서 record로 변경: 이를 통해 코드의 간결성을 높였습니다.
